### PR TITLE
Gradient norm clipping.

### PR DESCRIPTION
**Pull request by tdomhan**
#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md

### CODE REVIEW

File: sockeye/arguments.py [hunk 1]
```diff
@@ -700,6 +700,10 @@ def add_training_args(params):
                               default=1.0,
                               help='Clip absolute gradients values greater than this value. '
                                    'Set to negative to disable. Default: %(default)s.')
+    train_params.add_argument('--gradient-clipping-type',
```
Comment (fhieber) [inline sockeye/arguments.py]: can we rename to --clip-gradient-type to be consistent with the argument above?

File: sockeye/train.py [hunk 2]
```diff
@@ -555,9 +555,17 @@ def define_optimizer(args, lr_scheduler_instance) -> Tuple[str, Dict, str]:
                         "learning_rate": args.initial_learning_rate}
     if lr_scheduler_instance is not None:
         optimizer_params["lr_scheduler"] = lr_scheduler_instance
-    clip_gradient = none_if_negative(args.clip_gradient)
-    if clip_gradient is not None:
-        optimizer_params["clip_gradient"] = clip_gradient
+    gradient_clipping_threshold = none_if_negative(args.clip_gradient)
+    if gradient_clipping_threshold is None:
+        logger.info("Gradient clipping threshold set to negative value. Will not perform gradient clipping.")
+        gradient_clipping_type = "none"
```
Comment (fhieber) [inline sockeye/train.py]: why not use the constant defined in constants.py?

File: sockeye/train.py [hunk 3]
```diff
@@ -555,9 +555,17 @@ def define_optimizer(args, lr_scheduler_instance) -> Tuple[str, Dict, str]:
...
+    else:
+        gradient_clipping_type = args.gradient_clipping_type
+
+    # Note: for 'abs' we use the implementation inside of MXNet's optimizer and 'norm_*' we implement ourselves
+    # inside the TrainingModel.
+    if gradient_clipping_threshold is not None and gradient_clipping_type == "abs":
```
Comment (fhieber) [inline sockeye/train.py]: same here

File: sockeye/training.py [hunk 4]
```diff
@@ -40,6 +41,16 @@
 logger = logging.getLogger(__name__)
 
 
+def global_norm(ndarrays):
+    # accumulate in a list, as asscalar is blocking and this way we can run the norm calculation in parallel.
+    norms = []
+    for arr in ndarrays:
+        if arr is None:
+            continue
+        norms.append(mx.nd.square(mx.nd.norm(arr)))
```
Comment (fhieber) [inline sockeye/training.py]: would the parallelism be broken if a list comprehension would be used?

File: sockeye/training.py [hunk 5]
```diff
@@ -40,6 +41,16 @@
 logger = logging.getLogger(__name__)
 
 
+def global_norm(ndarrays):
```
Comment (fhieber) [inline sockeye/training.py]: can you type this method?

File: sockeye/training.py [hunk 6]
```diff
@@ -340,7 +371,7 @@ def _fit(self,
         :param lr_decay_opt_states_reset: How to reset optimizer states after learning rate decay.
         """
         # TODO: Push update to MXNet to expose the optimizer (Module should have a get_optimizer method)
```
Comment (fhieber) [inline sockeye/training.py]: can this TODO be moved up to the new _get_optimizer() method?

File: sockeye/training.py [hunk 7]
```diff
@@ -698,6 +733,22 @@ def load_checkpoint(self, directory: str, train_iter: data_io.ParallelBucketSent
         # And our own state
         return self.load_state(os.path.join(directory, C.TRAINING_STATE_NAME))
 
+    def get_global_grad_norm(self):
```
Comment (fhieber) [inline sockeye/training.py]: add return type

File: sockeye/training.py [hunk 8]
```diff
@@ -698,6 +733,22 @@ def load_checkpoint(self, directory: str, train_iter: data_io.ParallelBucketSent
...
+        executors = self._get_executors()
+        rescale_grad = self._get_optimizer().rescale_grad
+        # average norm across executors:
+        exec_norms = [global_norm([arr for arr in exe.grad_arrays if arr is not None]) for exe in executors]
+        norm_val = sum(exec_norms) / float(len(exec_norms))
+        norm_val *= rescale_grad
+        return norm_val
+
+    def rescale_grad(self, scale):
```
Comment (fhieber) [inline sockeye/training.py]: add typing

**Comment by fhieber** [comment]
It might also make sense to add norm clipping to one of the system tests.

File: sockeye/arguments.py (hunk 1 above)
Comment (tdomhan) [reply sockeye/arguments.py]: mh sure. or rename the other arg to `--gradient-clipping-treshold`, that even how it's called in train.py already.

File: sockeye/train.py (hunk 2 above)
Comment (tdomhan) [reply sockeye/train.py]: good point. I missed this line.

File: sockeye/train.py (hunk 3 above)
Comment (tdomhan) [reply sockeye/train.py]: changed.

File: sockeye/training.py (hunk 5 above)
Comment (tdomhan) [reply sockeye/training.py]: sure, done.

File: sockeye/training.py (hunk 4 above)
Comment (tdomhan) [reply sockeye/training.py]: I don't think so. I guess it makes sense to keep asscalar and the rest apart. I'll change it to a list comprehension.

File: sockeye/training.py (hunk 6 above)
Comment (tdomhan) [reply sockeye/training.py]: makes sense.

**Comment by tdomhan** [comment]
Addressed comments, added to one of the system tests and rebased.

**Comment by fhieber** [comment]
Thanks for iterating, I will merge this despite the CNN system test fail (which seems to be occuring independent of this PR).
